noir-projects/noir-contracts/contracts/token_contract/src/main.nr

```diff
@@ -613,64 +616,44 @@ contract Token {
-    ///
-    /// Note 1: fee_payer_npk is part of the fee_payer address preimage derivation, and is assumed to be known. So
-    //          if we have a known set of fee payer contract addresses getting fee_payer_npk and fee_payer_slot is
-    //          trivial (slot is derived in a `Map<...>` as a hash of balances map slot and a fee payer address).
-    /// Note 2: fee_payer_point and user_point above are public information because they are passed as args to
-    ///         the public `complete_refund(...)` function.
     // docs:start:setup_refund
+    /// Called by fee payer contract (FPC) to set up a refund for a user during the private fee payment flow.
     #[private]
     fn setup_refund(
-        fee_payer: AztecAddress, // Address of the entity which will receive the fee note.
         user: AztecAddress, // A user for which we are setting up the fee refund.
-        funded_amount: Field, // The amount the user funded the fee payer with (represents fee limit).
+        max_fee: Field, // The maximum fee a user is willing to pay for the tx (denominated in FPC's accepted asset)
         nonce: Field, // A nonce to make authwitness unique.
     ) {
-        // 1. This function is called by fee paying contract (fee_payer) when setting up a refund so we need to support
-        // the authwit flow here and check that the user really permitted fee_payer to set up a refund on their behalf.
+        // 1. This function is called by FPC when setting up a refund so we need to support the authwit flow here
+        // and check that the user really permitted fee_recipient to set up a refund on their behalf.
         assert_current_call_valid_authwit(&mut context, user);
 
-        // 3. Deduct the funded amount from the user's balance - this is a maximum fee a user is willing to pay
-        // (called fee limit in aztec spec). The difference between fee limit and the actual tx fee will be refunded
-        // to the user in the `complete_refund(...)` function.
+        // 2. Deduct the max fee from user's balance. The difference between max fee and the actual tx fee will
+        // be refunded to the user in the `complete_refund(...)` function.
         let change = subtract_balance(
             &mut context,
             storage,
             user,
-            U128::from_integer(funded_amount),
+            U128::from_integer(max_fee),
             INITIAL_TRANSFER_CALL_MAX_NOTES,
         );
+        // Emit the change note.
         storage.balances.at(user).add(user, change).emit(encode_and_encrypt_note_unconstrained(
             &mut context,
             user,
             user,
         ));
 
-        // 4. We prepare the partial notes
-        let fee_payer_point_slot =
-            _prepare_private_balance_increase(user, fee_payer, &mut context, storage);
+        // 3. Prepare the partial note for the refund.
         let user_point_slot = _prepare_private_balance_increase(user, user, &mut context, storage);
```

Comment:
If the "teardown phase" reverts, will we end up with a dangling partial note that can never be completed? Or is there a possible vulnerability whereby the user could complete that dangling partial note in a later tx?

Reply:
That's correct, the partial note would end up being stored in the public state and it could be considered a finalized later on. But it's not a vulnerability in this case because the [prepare_private_balance_increase(to, from)](prepare_private_balance_increase) can be called by anyone and the contract is designed in such a way that it's fine.